# 3번 문제 : 생일이 느린 순서로 10명 출력

In [9]:
from google.colab import files
uploaded = files.upload()

import pandas as pd

class Heap:
    def __init__(self, *args):
        self.__A = args[0] if args else []

    def insert(self, x):
        self.__A.append(x)
        self.__percolateUp(len(self.__A) - 1)

    def __percolateUp(self, i):
        parent = (i - 1) // 2
        if i > 0 and self.__A[i] > self.__A[parent]:
            self.__A[i], self.__A[parent] = self.__A[parent], self.__A[i]
            self.__percolateUp(parent)

    def deleteMax(self):
        if not self.isEmpty():
            max = self.__A[0]
            self.__A[0] = self.__A.pop()
            self.__percolateDown(0)
            return max
        return None

    def __percolateDown(self, i):
        child = 2 * i + 1
        right = 2 * i + 2
        if child <= len(self.__A) - 1:
            if right <= len(self.__A) - 1 and self.__A[child] < self.__A[right]:
                child = right
            if self.__A[i] < self.__A[child]:
                self.__A[i], self.__A[child] = self.__A[child], self.__A[i]
                self.__percolateDown(child)

    def max(self):
        return self.__A[0]

    def buildHeap(self):
        for i in range((len(self.__A) - 2) // 2, -1, -1):
            self.__percolateDown(i)

    def isEmpty(self):
        return len(self.__A) == 0

    def clear(self):
        self.__A = []

    def size(self):
        return len(self.__A)

file_path = 'DS_Birthdaydata.csv'
df = pd.read_csv(file_path)

heap = Heap()

for _, row in df.iterrows():
    raw_birthday = row['생년월일8자리(예.20040101)']
    if pd.isna(raw_birthday) or raw_birthday == '':
        continue
    birthday = str(int(raw_birthday))
    name = row['이름']
    student_id = row['학번']
    heap.insert((birthday, name, student_id))

print("생일이 가장 느린 사람 10명:")
for _ in range(10):
    if not heap.isEmpty():
        birthday, name, student_id = heap.deleteMax()
        print(f"{student_id} - {name} - {birthday}")


Saving DS_Birthdaydata.csv to DS_Birthdaydata.csv
생일이 가장 느린 사람 10명:
******82 - 홍서연 - 20241282
******22 - 신수민 - 20051230
******42 - 이서영 - 20051225
******69 - 강민주 - 20051214
******78 - 김민경 - 20051202
******41 - 이서영 - 20051112
******17 - 배시은 - 20051102
******87 - 김여원 - 20051031
******44 - 이서진 - 20051028
******64 - 서홍빈 - 20051024


DS_Birthdaydata.csv
DS_Birthdaydata.csv(text/csv) - 3908 bytes, last modified: 2025. 4. 14. - 100% done
Saving DS_Birthdaydata.csv to DS_Birthdaydata.csv
생일이 가장 느린 사람 10명:
******82 - 홍서연 - 20241282
******22 - 신수민 - 20051230
******42 - 이서영 - 20051225
******69 - 강민주 - 20051214
******78 - 김민경 - 20051202
******41 - 이서영 - 20051112
******17 - 배시은 - 20051102
******87 - 김여원 - 20051031
******44 - 이서진 - 20051028
******64 - 서홍빈 - 20051024

# 4번 문제 : 조원 이름과 생년월일 출력

In [17]:
class BidirectNode:
    def __init__(self, item, prev, next):
        self.item = item
        self.prev = prev
        self.next = next

class CircularDoublyLinkedList:
	def __init__(self):
		self.__head = BidirectNode("dummy", None, None)
		self.__head.prev = self.__head
		self.__head.next = self.__head
		self.__numItems = 0

	def insert(self, i:int, newItem) -> None:
		if (i >= 0 and i <= self.__numItems):
			prev = self.getNode(i - 1)
			newNode = BidirectNode(newItem, prev, prev.next)
			newNode.next.prev = newNode
			prev.next = newNode
			self.__numItems += 1
		else:
			print("index", i, ": out of bound in insert()") # 필요 시 에러 처리

	def append(self, newItem) -> None:
		prev = self.__head.prev
		newNode = BidirectNode(newItem, prev, self.__head)
		prev.next = newNode
		self.__head.prev = newNode
		self.__numItems += 1

	def pop(self, *args):
		# 가변 파라미터. 인자가 없거나 -1이면 마지막 원소로 처리하기 위함. 파이썬 리스트 규칙 만족
		if self.isEmpty():
			return None
		# 인덱스 i 결정
		if len(args) != 0: # pop(k)과 같이 인자가 있으면 i = k 할당
			i = args[0]
		if len(args) == 0 or i == -1:# pop()에 인자가 없거나 pop(-1)이면 i에 맨 끝 인덱스 할당
			i = self.__numItems - 1
		# i번 원소 삭제
		if (i >= 0 and i <= self.__numItems - 1):
			curr = self.getNode(i)
			retItem = curr.item
			curr.prev.next = curr.next
			curr.next.prev = curr.prev
			self.__numItems -= 1
			return retItem
		else:
			return None

	def get(self, *args):
		# 가변 파라미터. 인자가 없거나 -1이면 마지막 원소로 처리하기 위함. 파이썬 리스트 규칙 만족
		if self.isEmpty(): return None
		# 인덱스 i 결정
		if len(args) != 0:   # pop(k)과 같이 인자가 있으면 i = k 할당
			i = args[0]
		if len(args) == 0 or i == -1:
			i = self.__numItems - 1
		# i번 원소 리턴
		if (i >= 0 and i <= self.__numItems - 1):
			return self.getNode(i).item
		else:
			return None

	def isEmpty(self) -> bool:
		return self.__numItems == 0

	def __findNode(self, x) -> BidirectNode:
		curr = self.__head.next  # 0번 노드
		while curr != self.__head:

			if curr.item == x:
				return curr
			else:
				curr = curr.next
		return None

	def getNode(self, i:int) -> BidirectNode:
		curr = self.__head  # 더미 헤드, index: -1
		for index in range(i + 1):
			curr = curr.next
		return curr

class Friend:
    def __init__(self, name, birth_date):
        self.name = name
        self.birth_date = birth_date

from google.colab import drive
drive.mount('/content/drive')

import pandas as pd
file_path = '/content/drive/MyDrive/DS_Birthdaydata.csv'
df = pd.read_csv(file_path)

circular_doubly_linked_list = CircularDoublyLinkedList()

for _, row in df.iterrows():
    name = row['이름']
    raw_birthday = row['생년월일8자리(예.20040101)']

    if pd.isna(raw_birthday) or raw_birthday == '':
        continue

    birthday = str(int(raw_birthday))
    friend = Friend(name, birthday)
    circular_doubly_linked_list.append(friend)

my_team_names = ["김연진","변수연","정윤서","박서연","노은서","오세은","박성연","김민경","김보민","홍서연"]


curr = circular_doubly_linked_list._CircularDoublyLinkedList__head.next
while curr != circular_doubly_linked_list._CircularDoublyLinkedList__head:
    if curr.item.name in my_team_names:
        print(f"이름: {curr.item.name}, 생년월일: {curr.item.birth_date}")
    curr = curr.next

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
이름: 김민경, 생년월일: 20051202
이름: 김보민, 생년월일: 20020911
이름: 김연진, 생년월일: 20010826
이름: 노은서, 생년월일: 20050316
이름: 박서연, 생년월일: 20040428
이름: 박성연, 생년월일: 20040514
이름: 변수연, 생년월일: 20040802
이름: 오세은, 생년월일: 20050328
이름: 정윤서, 생년월일: 20030802
이름: 홍서연, 생년월일: 20040611
이름: 홍서연, 생년월일: 20241282


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
이름: 김민경, 생년월일: 20051202
이름: 김보민, 생년월일: 20020911
이름: 김연진, 생년월일: 20010826
이름: 노은서, 생년월일: 20050316
이름: 박서연, 생년월일: 20040428
이름: 박성연, 생년월일: 20040514
이름: 변수연, 생년월일: 20040802
이름: 오세은, 생년월일: 20050328
이름: 정윤서, 생년월일: 20030802
이름: 홍서연, 생년월일: 20040611
이름: 홍서연, 생년월일: 20241282

# 교재 8장 연습문제

연습문제 1:

가질 수 있다.
루트 노드로부터 다른 가지에서 나온 경우 임의의 최대 힙에서 더 얕은 곳에 있는 원소가 더 깊은 곳에 있는 원소보다 더 작은 값을 가질 수 있다.

연습문제 2:

항상 가장 작은 값을 가지는 것은 아니다.
A[n-1]은 부모노드보다 작은 값을 가지는 것은 맞지만 전체 노드에서 가장 작은 값을 보장하지는 않는다.

연습문제 3:

n-1-((n-2)//2)
스며내리기는 자식 노드가 있는 노드만 수행한다. 따라서 스며내리기를 하는 노드 수는 (n-2) //2 + 1이고, 스며내리기를 하지 않는 노드의 수는 n-1 - ((n-2)//2) 이다

연습문제 4:

최악의 경우 : Θ(logn)
최선의 경우 : Θ(1)

최악의 경우는 가장 아래까지 내려가는경우 즉, 트리의 높이인 Θ(logn)의 경우이고, 최선의 경우는 교환없이 종료되는 Θ(1)의 경우이다.

연습문제 5:

간단하다.
테스트용 마지막 요소 삭제는 pop으로 가능하고 이는 힙 성질에 영향을 주지 않기에 간단하다고 판단할 수 있다.


연습문제 6:

비효율적이며, Θ(nlogn)의 시간이 소요된다.
스며내리기의 경우는 최악의 경우 Θ(n)의 시간이 걸린다. 스며오르기의 경우에는 깊이가 깊어질수록 비교와 교체 작업이 많아져서 n/2 * logn + n/4 * log(n-1) .. 즉, 	Θ(nlogn)의 시간이 소요되고,비효율적이다.

연습문제 7:

O(logn)
바뀐 노드를 삭제하고(O(logn)) 바꾼 값을 다시 삽입(O(logn))한다. 즉, 시간복잡도는 O(logn)이다.

# LeetCode 703.Kth Largest Element in Stream

In [6]:
import heapq

class KthLargest:
    def __init__(self, k, nums):
        self.k = k
        self.heap = nums
        heapq.heapify(self.heap)
        while len(self.heap) > k:
            heapq.heappop(self.heap)

    def add(self, val):
        heapq.heappush(self.heap, val)
        if len(self.heap) > self.k:
            heapq.heappop(self.heap)
        return self.heap[0]

if __name__ == "__main__":
    commands = ["KthLargest", "add", "add", "add", "add", "add"]
    print(commands)
    inputs = [[3, [4, 5, 8, 2]], [3], [5], [10], [9], [4]]
    print(inputs)

    output = []
    kth = None

    for cmd, inp in zip(commands, inputs):
        if cmd == "KthLargest":
            k, nums = inp
            kth = KthLargest(k, nums)
            output.append(None)
        elif cmd == "add":
            val = inp[0]
            result = kth.add(val)
            output.append(result)

    print(output)

['KthLargest', 'add', 'add', 'add', 'add', 'add']
[[3, [4, 5, 8, 2]], [3], [5], [10], [9], [4]]
[None, 4, 5, 5, 8, 8]
